### PR TITLE
Point to variable in `asm!` macro when failing borrowck

### LINKED ISSUE #34940: Span for asm error too large
Looking around the compile-fail test suite, I noticed this example:

``` term
jturner-23759:rust jturner$ ./build/x86_64-apple-darwin/stage1/bin/rustc src/test/compile-fail/asm-out-assign-imm.rs 
error[E0384]: re-assignment of immutable variable `x`
  --> src/test/compile-fail/asm-out-assign-imm.rs:24:9
   |
21 |     x = 1; //~ NOTE first assignment
   |     ----- first assignment to `x`
...
24 |         asm!("mov $1, $0" : "=r"(x) : "r"(5));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ re-assignment of immutable variable
src/test/compile-fail/asm-out-assign-imm.rs:24:9: 24:47 note: in this expansion of asm!
```

The span on the asm line seems a bit much.  Ideally, I think we'd show them where the assignment is happening so they don't have to look for it.  Something like:

``` term
24 |         asm!("mov $1, $0" : "=r"(x) : "r"(5));
   |                             ^^^^^^^ re-assignment of immutable variable
src/test/compile-fail/asm-out-assign-imm.rs:24:9: 24:47 note: in this expansion of asm!
```

Or even:

``` term
24 |         asm!("mov $1, $0" : "=r"(x) : "r"(5));
   |                                  ^ re-assignment of immutable variable
src/test/compile-fail/asm-out-assign-imm.rs:24:9: 24:47 note: in this expansion of asm!
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #34940.

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -1812,6 +1812,7 @@ pub struct InlineAsmOutput {
     pub constraint: Symbol,
     pub is_rw: bool,
     pub is_indirect: bool,
+    pub span: Span,
```
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]: Looks like there's no need to keep span here additionally, output expressions are already preserved in `hir::ExprKind::InlineAsm` (the first `HirVec<Expr>`), so spans can be taken from there.

Comment (estebank) [reply src/librustc/hir/mod.rs]: This addition is needed in https://github.com/rust-lang/rust/pull/54945/files#diff-5b4d01d26caf43976125ba0f877e78c0 (for the nll output) as we have `StatementKind::InlineAsm` available only. `ExprKind::InlineAsm` is accessible for ast-based borrow checker and there it is indeed using the `expr.span` (https://github.com/rust-lang/rust/pull/54945/files#diff-dc986596b3470440badc6fb1a667adca).

Comment (petrochenkov) [reply src/librustc/hir/mod.rs]: I see.

**Pull request merged**